### PR TITLE
[10.0][CHG] partner_address_version: refactor the way versioning is managed

### PR DESCRIPTION
**Pull request by Cedric-Pigeon**

### CODE REVIEW

**Comment by sebastienbeau** [comment]
@bguillot maybe you can be interested

File: partner_address_version/models/res_partner.py [hunk 1]
```diff
@@ -52,15 +31,50 @@ def get_version_hash(self):
         version_hash = hashlib.md5(str(version)).hexdigest()
         return version_hash
 
+    def _get_impacted_tables(self):
+        return ['sale_order', 'sale_order_line', 'account_invoice',
```
Comment (Cedric-Pigeon) [inline partner_address_version/models/res_partner.py]: this line will be moved to sale_partner_version

File: partner_address_version/models/res_partner.py [hunk 2]
```diff
@@ -52,15 +31,54 @@ def get_version_hash(self):
         version_hash = hashlib.md5(str(version)).hexdigest()
         return version_hash
 
+    def _get_impacted_tables(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py]:
```suggestion
    def _version_impacted_tables(self):
```

File: partner_address_version/models/res_partner.py [hunk 3]
```diff
@@ -52,15 +31,54 @@ def get_version_hash(self):
...
+        """
+        :return:
+            - list of tables to update in case of address versioning
+        """
+        return []
+
+    def manage_versionning(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py]:
```suggestion
    def _version_apply(self):
```

File: partner_address_version/models/res_partner.py [hunk 4]
```diff
@@ -13,27 +13,6 @@ class ResPartner(models.Model):
...
-        default = {
-            'active': False,
-            'version_hash': version_hash,
-            'parent_id': self.parent_id and self.parent_id.id or self.id,
-            'date_version': fields.Datetime.now(),
-        }
-        versioned_partner = self.copy(default=default)
-        return versioned_partner
-
     def get_version_fields(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py:16]:
```suggestion
    def _version_fields(self):
```

File: partner_address_version/models/res_partner.py [hunk 5]
```diff
@@ -52,15 +31,54 @@ def get_version_hash(self):
...
+            if not partner.version_hash and not vals.get('version_hash', False):
+                partner.manage_versionning()
+
             if partner.version_hash and has_written_versioned_fields:
                 raise exceptions.UserError(_(
                     "You can't modify a versioned field %s on the versioned "
                     "partner %s.") % (version_fields, partner.name))
         return super(ResPartner, self).write(vals)
+
+    def _create_partner_version(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py]:
```suggestion
    def _version_create(self):
```

File: partner_address_version/models/res_partner.py [hunk 6]
```diff
@@ -52,15 +31,54 @@ def get_version_hash(self):
...
+        version_hash = self.get_version_hash()
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py]:
```suggestion
        version_hash = self._version_hash()
```

File: partner_address_version/models/res_partner.py [hunk 7]
```diff
@@ -52,15 +31,54 @@ def get_version_hash(self):
...
+        default = {
+            'active': False,
+            'version_hash': version_hash,
+            'parent_id': self.parent_id and self.parent_id.id or self.id,
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py]:
```suggestion
            'parent_id': self.parent_id.id if self.parent_id else self.id,
```

File: partner_address_version/wizards/base_partner_merge.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+
+
+class MergePartnerAutomatic(models.TransientModel):
+    """
```
Comment (simahawk) [inline partner_address_version/wizards/base_partner_merge.py]: is this docstring custom?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: partner_address_version/wizards/base_partner_merge.py (hunk 8 above)
Comment (Cedric-Pigeon) [reply partner_address_version/wizards/base_partner_merge.py]: bad copy/paste ...

**Comment by Cedric-Pigeon** [comment]
@simahawk thanks for your review. comments attended.

File: partner_address_version/models/res_partner.py [hunk 9]
```diff
@@ -13,54 +13,69 @@ class ResPartner(models.Model):
...
-        default = {
-            'active': False,
-            'version_hash': version_hash,
-            'parent_id': self.parent_id and self.parent_id.id or self.id,
-            'date_version': fields.Datetime.now(),
-        }
-        versioned_partner = self.copy(default=default)
-        return versioned_partner
-
-    def get_version_fields(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py:37]: maybe worth to keep the methods that you've renamed and log deprecated?

File: partner_address_version/tests/test_address_version.py [hunk 10]
```diff
@@ -33,25 +33,19 @@ def test_hash(self):
         self.assertEqual(test_hash, self.partner.get_version_hash())
 
     def test_create_version_partner(self):
-        new_partner = self.partner.get_address_version()
+        new_partner = self.partner._create_partner_version()
```
Comment (simahawk) [inline partner_address_version/tests/test_address_version.py]: isn't this `_version_create` now?

File: partner_address_version/models/res_partner.py [hunk 11]
```diff
@@ -13,54 +13,69 @@ class ResPartner(models.Model):
...
         return version_hash
 
+    def _version_impacted_tables(self):
+        """
+        :return:
+            - list of tables to update in case of address versioning
+        """
+        return []
+
+    def _version_need(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py:59]: hmm this is wanted? `_version_apply` won't run...

Comment (Cedric-Pigeon) [reply partner_address_version/models/res_partner.py:59]: this is supposed to be overriden.
See https://github.com/OCA/sale-workflow/pull/1119/files#diff-9631fa4efa5849317b2d7720e17c8a75R11

Comment (simahawk) [reply partner_address_version/models/res_partner.py:59]: then please add a docstring ;)

Comment (Cedric-Pigeon) [reply partner_address_version/models/res_partner.py:59]: indeed, done ;-)

Comment (Cedric-Pigeon) [reply partner_address_version/models/res_partner.py:59]: indeed, done ;-)

**Comment by rousseldenis** [comment]
@simahawk Could you update your review ?

File: partner_address_version/models/res_partner.py [hunk 12]
```diff
@@ -13,54 +13,93 @@ class ResPartner(models.Model):
...
+
+    def _version_fields(self):
         return [
             'name', 'street', 'street2', 'zip', 'city', 'country_id',
             'parent_id'
         ]
 
     def get_version_hash(self):
-        version_fields = self.get_version_fields()
+        # deprecated uses _version_hash instead
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py:27]: if you want to support backward compat  you should log a deprecation warning :wink:

File: partner_address_version/models/res_partner.py [hunk 13]
```diff
@@ -13,54 +13,93 @@ class ResPartner(models.Model):
     version_hash = fields.Char(readonly=True, copy=False)
     date_version = fields.Datetime(string='Date version', readonly=True)
 
-    @api.multi
-    def get_address_version(self):
```
Comment (simahawk) [inline partner_address_version/models/res_partner.py:17]: as you have done w/ `get_version_fields`, can we have a deprecation warning?

**Review by simahawk — COMMENTED** [review]
@Cedric-Pigeon sorry, I missed further updates :sweat_smile:

File: partner_address_version/__manifest__.py [hunk 14]
```diff
@@ -12,5 +12,6 @@
     'installable': True,
     'depends': [
         'base',
```
Comment (NL66278) [inline partner_address_version/__manifest__.py:14]: You can leave out base then.

**Comment by NL66278** [comment]
@Cedric-Pigeon Will you be taking up last comments? Would be nice to merge...

**Comment by NL66278** [comment]
I will merge anyway as comments are not critical

**Review by NL66278 — APPROVED** [review]
:+1: Minor comments, but not enough to block these changes.

**Pull request merged**